### ISSUE TITLE
ReadOnlyDictionary Keys/Values property does not reflect changes if underlying Dictionary is a ConcurrentDictionary and was used one time before

### ISSUE THREAD

**Issue by akacoding** [dotnet/runtime#93829]
### Description

If exposing a `ConcurrentDictionary` readonly with `ReadOnlyDictionary` and using the `Keys` or `Values` property of the `ReadOnlyDictionary` it will not reflect any changes made to the underlying Dictionary after using `Keys` of `Values` of the `ReadOnlyDictionary` one time before.

That is because the `ConcurrentDictionary.Keys`/`ConcurrentDictionary.Values` creates a snapshot and the `ReadOnlyDictionary` caches it to a field. This cached snapshot never changes again after retrieved once.

If used with a regular `Dictionary` the `Keys`/`Values` property of `ReadOnlyDictionary` works fine.

I've tested it for `.Net 4.7.2` and `.Net 8 Preview 3`. In both cases it behaves like written above.

### Reproduction Steps

    using System;
    using System.Collections.Concurrent;
    using System.Collections.ObjectModel;
                        
    public class Program
    {
        public static void Main()
        {            
            var concurrentDictionary = new ConcurrentDictionary<string, string>();
            concurrentDictionary.TryAdd("Key_A", "Value_A");
            concurrentDictionary.TryAdd("Key_B", "Value_B");
            
            var concurrentDictionaryReadOnly = new ReadOnlyDictionary<string, string>(concurrentDictionary);
            
            Console.WriteLine($"ConcurrentDictionary count: {concurrentDictionaryReadOnly.Count}");
            
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary");
            foreach (var kvp in concurrentDictionaryReadOnly)
                Console.WriteLine($"Key: {kvp.Key}: /  Value: {kvp.Value}");
            
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary.Keys");
            foreach (var k in concurrentDictionaryReadOnly.Keys)
                Console.WriteLine($"Key: {k}");
            
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary.Values");
            foreach (var v in concurrentDictionaryReadOnly.Values)
                Console.WriteLine($"Value: {v}");
            
            Console.WriteLine();
            Console.WriteLine("Add another item to ConcurrentDictionary");
            concurrentDictionary.TryAdd("Another_Key", "Another_Value");
            
            Console.WriteLine();
            Console.WriteLine($"ConcurrentDictionary count: {concurrentDictionaryReadOnly.Count}");
            
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary");
            foreach (var kvp in concurrentDictionaryReadOnly)
                Console.WriteLine($"Key: {kvp.Key}: /  Value: {kvp.Value}");
            
            // The following Keys/Values calls will not contain the previous added item
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary.Keys");
            foreach (var k in concurrentDictionaryReadOnly.Keys)
                Console.WriteLine($"Key: {k}");
            
            Console.WriteLine();
            Console.WriteLine("Iterating ConcurrentDictionary.Values");
            foreach (var v in concurrentDictionaryReadOnly.Values)
                Console.WriteLine($"Value: {v}");
        }
    }

The output will be the following:

> ConcurrentDictionary count: 2
> 
> Iterating ConcurrentDictionary
> Key: Key_B: /  Value: Value_B
> Key: Key_A: /  Value: Value_A
> 
> Iterating ConcurrentDictionary.Keys
> Key: Key_B
> Key: Key_A
> 
> Iterating ConcurrentDictionary.Values
> Value: Value_B
> Value: Value_A
> 
> Add another item to ConcurrentDictionary
> 
> ConcurrentDictionary count: 3
> 
> Iterating ConcurrentDictionary
> Key: Another_Key: /  Value: Another_Value
> Key: Key_B: /  Value: Value_B
> Key: Key_A: /  Value: Value_A
> 
> Iterating ConcurrentDictionary.Keys
> Key: Key_B
> Key: Key_A
> 
> Iterating ConcurrentDictionary.Values
> Value: Value_B
> Value: Value_A

### Expected behavior

The `ReadOnlyDictionary,Keys` and `ReadOnlyDictionary,Values` properties should reflect changes to the underlying `ConcurrentDictionary`. In case of a regular `Dictionary` it already works that way. 

### Actual behavior

The `ReadOnlyDictionary,Keys` and `ReadOnlyDictionary,Values` properties does not reflect changes to the underlying `ConcurrentDictionary` if used once before.

### Regression?

No

**Comment by colejohnson66** [comment]
I would say this is "by design". For a concurrent object to exist in a language without move semantics (i.e. basically any popular language except Rust), it _must_ make a copy of the input sequence. Otherwise, it has no guarantee that it is the only one modifying the underlying collection.

Being able to modify a concurrent dictionary out from underneath it (through the input collection) would violate this, and make it no safer than the normal, non-concurrent, type.

**Comment by skyoxZ** [comment]
```csharp
Dictionary<string, string> dict = new();
var keys = dict.Keys;
dict.Add("a", "");
Console.WriteLine(keys.Count);
```
The code above returns `1` and this behavior is well-defined:
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.keys?view=net-7.0#remarks
> The returned Dictionary<TKey,TValue>.KeyCollection is not a static copy; instead, the Dictionary<TKey,TValue>.KeyCollection refers back to the keys in the original Dictionary<TKey,TValue>. Therefore, changes to the Dictionary<TKey,TValue> continue to be reflected in the Dictionary<TKey,TValue>.KeyCollection.

However, `ConcurrentDictionary.Keys` returns a static copy: ([seems not documented](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.keys?view=net-7.0))
```csharp
ConcurrentDictionary<string, string> cd = new();
var ckeys = cd.Keys;
cd.TryAdd("a", "");
Console.WriteLine(ckeys.Count); // print 0
```
I don't think it's a bug. And the behavior is unlikely to be changed. 



Anyway, we should fix `ReadOnlyDictionary,Keys` for no guarantee that the underlying `IDictionary`'s `Keys` returns a dynamic collection.
Similar situation on `Values`.

**Comment by eiriktsarpalis** [comment]
I've opened a PR that should address the issue: https://github.com/dotnet/runtime/pull/93870/

**Comment by akacoding** [comment]
> However, `ConcurrentDictionary.Keys` returns a static copy: ([seems not documented](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.keys?view=net-7.0))
> 
> ```cs
> ConcurrentDictionary<string, string> cd = new();
> var ckeys = cd.Keys;
> cd.TryAdd("a", "");
> Console.WriteLine(ckeys.Count); // print 0
> ```
> 
> I don't think it's a bug. And the behavior is unlikely to be changed.

Yes, it seems not documented. In the case of [ConcurrentDictionary.Count](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.count?view=net-7.0) it is documented that it uses snapshot semantics.

It is also my opinion that the `ConcurrentDictionary` behaves correctly. The snapshot behavior (and also acquiring all locks) of `Keys`, `Values` and `Count` are also mentioned by some blog posts.

> Anyway, we should fix `ReadOnlyDictionary.Keys` for no guarantee that the underlying `IDictionary`'s `Keys` returns a dynamic collection.
> Similar situation on `Values`.

Yes, that seems to be the solution. But we should think about performance when using the returned `KeyCollection` for more than just iteration with one call to `GetEnumerator`. If so it would result in populating a new snapshot of the `ConcurrentDictionary` all the time. It should be rare that it is used more than that but we cannot know,
